### PR TITLE
C++: Support repeated initializers in dbscheme.

### PR DESCRIPTION
**Pull request by sashabu**
I haven't added a change note as this PR doesn't change the QL API, only the database schema.

### CODE REVIEW

**Comment by MathiasVP** [comment]
We'll also need to run DCA on this PR to ensure that the updated stats file doesn't introduce any bad joins. If you do the DCA run I'll be happy to help you 👀 it (and also help fix whatever needs to be fixed). And I'm sure @jketema will be able to help with this as well.

**Comment by jketema** [comment]
> We'll also need to run DCA on this PR to ensure that the updated stats file doesn't introduce any bad joins. If you do the DCA run I'll be happy to help you 👀 it (and also help fix whatever needs to be fixed). And I'm sure @jketema will be able to help with this as well.

I think that should happen on the internal PR. Which should bump the submodule (it might do that, I haven't checked yet).

File: cpp/ql/lib/upgrades/a5bb28ed29f73855d64cc5f939cef977fa8fd19a/upgrade.properties [hunk 1]
```diff
@@ -0,0 +1,4 @@
+description: Support repeated initializers, which are allowed in C with designated initializers.
+compatibility: full
```
Comment (jketema) [inline cpp/ql/lib/upgrades/a5bb28ed29f73855d64cc5f939cef977fa8fd19a/upgrade.properties:2]: Shouldn't this be backwards? Or will the old database also contain all the tuples for repeated initialisers (it's not clear to me what the database will look like when there are dataset check errors)?

File: cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties [hunk 2]
```diff
@@ -0,0 +1,4 @@
+description: Revert support for repeated initializers, which are allowed in C with designated initializers.
+compatibility: backwards
```
Comment (jketema) [inline cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties:2]: What will the resulting downgraded database look like when a repeated initialiser occurs? With both tuples be present, and will the database just suffer from an inconsistency, or is one of the tuples dropped?

**Review by jketema — COMMENTED** [review]
Two small comments, otherwise this part LGTM.

File: cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties (hunk 2 above)
Comment (sashabu) [reply cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties:2]: Both tuples will be present and the database will suffer from an inconsistency.

File: cpp/ql/lib/upgrades/a5bb28ed29f73855d64cc5f939cef977fa8fd19a/upgrade.properties (hunk 1 above)
Comment (sashabu) [reply cpp/ql/lib/upgrades/a5bb28ed29f73855d64cc5f939cef977fa8fd19a/upgrade.properties:2]: The old database contains the repeated initializers (with dataset check errors). The upgrade script keeps and orders them correctly, so we should have exactly the same table as we'd get with the new extractor.

File: cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties (hunk 2 above)
Comment (sashabu) [reply cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties:2]: That's correct. I wasn't sure about how the compatibility categories apply to downgrade scripts. I'll change it to full.

Comment (jketema) [reply cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties:2]: > I wasn't sure about how the compatibility categories apply to downgrade scripts

You're basically reasoning backwards as if the new and updated extractor is producing the old database and the old extractor is producing our fresh and exciting updated database.

I must admit I must always pause for a moment and think about this.

Comment (sashabu) [reply cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties]: Done. I'm assuming we don't need to rerun tests for this?

Comment (jketema) [reply cpp/downgrades/19887dbd33327fb07d54251786e0cb2578539775/upgrade.properties]: No, I think that field is only informational.

**Review by jketema — APPROVED** [review]
LGTM.

**Pull request merged**